webhook_service/main.go

```diff
@@ -0,0 +1,873 @@
+
+	if resp.StatusCode == http.StatusNotFound {
+		ctx := context.Background()
+		str := fmt.Sprintf("%s:%s", message_404, *msg.Payload.MessageID)
+		rdb.Set(ctx, str, true, 0)
+		if msg.Payload.MessageID != nil {
+			return "", fmt.Errorf("HTTP 404 for %s", *msg.Payload.MessageID)
+		} else {
+			return "", fmt.Errorf("HTTP 404; no message ID included")
+		}
+	} else if resp.StatusCode == 400 || resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden {
+		return "", fmt.Errorf("%s is not a valid webhook URL", msg.Payload.URL)
+
+	} else if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusNoContent || resp.StatusCode == http.StatusTooManyRequests {
+		ClearWebhookError(rdb, webhookID)
+		// These responses will contain valid rate limit headers
+		bucket := resp.Header.Get("X-RateLimit-Bucket")
+		rateLimit.mu.Lock()
+		defer rateLimit.mu.Unlock()
+
+		if remaining, err := strconv.Atoi(resp.Header.Get("X-RateLimit-Remaining")); err == nil {
+			rateLimit.Remaining = remaining
+		}
+		if limit, err := strconv.Atoi(resp.Header.Get("X-RateLimit-Limit")); err == nil {
+			rateLimit.Limit = limit
+		}
+		if reset, err := strconv.ParseInt(resp.Header.Get("X-RateLimit-Reset"), 10, 64); err == nil {
+			rateLimit.ResetTime = time.Unix(reset, 0)
+			// Pad the reset time two seconds because Discord are liars and following their header
+			// responses exactly will still cause excessive 429s
+			rateLimit.ResetTime = rateLimit.ResetTime.Add(time.Duration(2.0 * float64(time.Second)))
+
+		}
+		if resetAfter, err := strconv.ParseFloat(resp.Header.Get("X-RateLimit-Reset-After"), 64); err == nil {
+			// Pad the reset time two seconds because Discord are liars and following their header
+			// responses exactly will still cause excessive 429s
+			rateLimit.ResetAfter = time.Duration((resetAfter + 2.0) * float64(time.Second))
+		}
+
+		if resp.StatusCode == http.StatusTooManyRequests {
+			UpdateBucketRateLimitCount(rdb, bucket)
+			logger.Warning(fmt.Sprintf("HTTP %d for %s", resp.StatusCode, bucket))
+			rateLimit.RateLimited = true
+			if resp.Header.Get("X-RateLimit-Global") != "" {
+				// Set the global rate limit flag in Redis so it can be retrieved by the API
+				// we don't explicitly check it because each worker will flag itself as rate limited
+				logger.Warning(fmt.Sprintf("Global rate limit hit, pausing for %v", rateLimit.ResetAfter))
+				SetGloballyRateLimited(rdb, rateLimit.ResetTime, rateLimit.ResetAfter)
+				// Force wait by maxing out the requests in our local window
+				CheckLocalRateLimit(localRateLimit, time.Now())
+			}
+			return bucket, fmt.Errorf("rate limited until %v", rateLimit.ResetTime)
+		}
+
+		rateLimit.RateLimited = false
+		return bucket, nil
+	} else {
+		return "", fmt.Errorf("unhandled HTTP status %d", resp.StatusCode)
+	}
+}
+
+// ProcessQueue handles a bucket queue with retries
+func (bw *BucketWorker) ProcessQueue(rdb *redis.Client) {
+	for {
+		bw.rateLimit.mu.Lock()
```

Comment:
I would always make a defer call to Unlock, just to make sure that all branches unlock the mutex. In this case, I think you covered this with explicit calls, but I would feel much safer still having this here :D

Reply:
After some reflection (and I think you mentioned it in a different comment?) I don't think we even need to rate limit here.

Each queue operates on a single bucket and shouldn't ever be competing for that struct.